zerver/actions/message_flags.py

```diff
@@ -43,13 +49,13 @@ def do_mark_all_as_read(user_profile: UserProfile) -> int:
     )
     do_clear_mobile_push_notifications_for_ids([user_profile.id], all_push_message_ids)
 
-    msgs = UserMessage.objects.filter(user_profile=user_profile).extra(
-        where=[UserMessage.where_unread()],
-    )
+    msgs = get_user_message_for_safe_bulk_update()
+    msgs = msgs.filter(user_profile=user_profile).extra(where=[UserMessage.where_unread()])
 
-    count = msgs.update(
-        flags=F("flags").bitor(UserMessage.flags.read),
-    )
+    with transaction.atomic():
```

Comment:
I think this pattern would be more readable if you put the definition inside the transaction block. For example (applying other comments):
```
with transaction.atomic(savepoint=False):
    query = UserMessage.select_for_update_query().filter(
        user_profile=user_profile
    ).extra(
        where=[UserMessage.where_unread()],
    )
   count = query.update(...)

```